### ISSUE TITLE
Audit the current Deep Sets point-set builder and propose stronger point features

### ISSUE THREAD

**Issue by annawoodard** [dsi-clinic/vanguard#119]
## Background
 
The Deep Sets path already exists, but it's intentionally minimal. Right now the builder keeps tumor-local skeleton points and stores only one point feature: `curvature_rad`. This subproject is about building **better point sets and better point features** — not changing the model architecture.
 
### What already exists
 
| File | Role |
|---|---|
| `build_deepsets_dataset.py` | Builds and saves one point set per case |
| `deepsets_data.py` | Data loading for training |
| `train_deepsets.py` | Deep Sets training entrypoint |
| `graph_extraction/README.md` | Documents the upstream centerline/graph outputs |
 
**Current inclusion rule:** keep skeleton points whose signed distance to the tumor is ≤ the local vessel radius. If nothing passes, fall back to the nearest 64 points.
 
**Current point features:** just `curvature_rad`.
 
**Full dataset builds should run through Slurm, not on the headnode.**
 
## Key Definitions
 
**Point** — One vessel/skeleton voxel included in a case's point set.
 
**Point feature** — A number attached to each point (e.g. curvature, vessel radius, signed distance to tumor).
 
**Inclusion rule** — The filter that decides which points make the cut (e.g. all points within 50 mm of the tumor, or the nearest N points).
 
**Manifest** — The case-level CSV that tells training where each saved point set lives.
 
### 1. Audit the current builder
 
Read `build_deepsets_dataset.py`, `deepsets_data.py`, and `train_deepsets.py`. Write a short summary of:
 
- What is currently stored per point
- What is stored per case in the manifest
- What metadata are available from the upstream centerline/graph outputs but currently dropped
 
### 2. Propose candidate point features
 
Make a plain list. For each candidate, note:
 
- **Feature name**
- **Short description**
- **Source files needed**
- **Whether it can be computed from existing saved outputs**
 
Here are strong candidates to consider:
 
- Signed distance to tumor boundary
- Absolute distance to tumor boundary
- Inside-tumor flag
- Shell bin or shell one-hot encoding
- Normalized XYZ offset from tumor centroid
- Local vessel radius from the support mask
- Node degree
- Endpoint / chain / bifurcation flags
- Local direction relative to the tumor centroid
- Point-local dynamic features sampled from the DCE time series
- Point-local graph-context features read from saved graph outputs
 
### 3. For dynamic point features, verify spatial alignment
 
If you propose features sampled from a DCE volume (or any other source volume), you need to confirm that the volume actually lines up with the saved vessel outputs for the same case. Check:
 
- Same case / study ID
- Same spatial grid (or a clearly compatible one)
- Matching array shape after loading
- Consistent spacing and orientation
 
#### Suggested alignment check visualization
 
Pick 2–3 representative cases and produce an **overlay figure** for each:
 
1. Load a single axial slice from the DCE source volume (e.g. the first post-contrast timepoint).
2. On the same slice, overlay the skeleton points as colored dots — color by an existing feature like `curvature_rad` or just by signed distance.
3. Also overlay the tumor mask contour.
 
If the skeleton points land on vessels visible in the DCE image and sit in the right spatial relationship to the tumor contour, alignment is trustworthy. If they're shifted, clipped, or floating in empty space, something is off.
 
A simple version in matplotlib:
 
```python
import matplotlib.pyplot as plt
import numpy as np
 
fig, ax = plt.subplots(1, 1, figsize=(8, 8))
 
# Show one axial slice from the DCE volume
ax.imshow(dce_slice, cmap="gray", origin="lower")
 
# Overlay tumor contour
ax.contour(tumor_mask_slice, levels=[0.5], colors="red", linewidths=1.5)
 
# Overlay skeleton points that fall on or near this slice
# (filter to points within ±1 voxel of the slice index)
ax.scatter(pts_col, pts_row, c=pts_feature, cmap="plasma", s=8, alpha=0.8)
 
ax.set_title(f"Case {case_id} — slice {slice_idx}")
plt.tight_layout()
plt.savefig(f"alignment_check_{case_id}.png", dpi=150)
```
 
Save these figures and include them in your design note. They serve as both a sanity check and documentation.
 
## Definition of Done
 
- A short note listing current point features and proposed next features
- Each proposed feature identifies what extra inputs it requires
- If dynamic point features are proposed, the note states whether alignment looks trustworthy, supported by overlay figures from 2–3 cases
 
## Helpful File Pointers
 
- `build_deepsets_dataset.py`
- `deepsets_data.py`
- `train_deepsets.py`
- `graph_extraction/README.md`

**Comment by jpivarski** [comment]
The following has been resolved by PR #129:

- [deepsets_data.py](/Users/jpivarski/dsi/vanguard/deepsets_data.py:42) to make `torch.load(..., weights_only=False)` conditional on whether that parameter exists, for PyTorch compatibility
- [deepsets_model.py](/Users/jpivarski/dsi/vanguard/deepsets_model.py:90) to add a fallback for max pooling when `Tensor.scatter_reduce_` is unavailable
- [slurm/deepsets_job.slurm](/Users/jpivarski/dsi/vanguard/slurm/deepsets_job.slurm:10) to drop `-u` from `set -euo pipefail`
- [slurm/submit_deepsets_pipeline.sh](/Users/jpivarski/dsi/vanguard/slurm/submit_deepsets_pipeline.sh:11) to normalize `OUT_ROOT` to an absolute path before submission

PR #129 improves old-PyTorch compatibility and Slurm submission robustness for the existing Deep Sets pipeline. But it does not touch the builder, add point features, audit upstream metadata, verify alignment, or produce the requested note/figures.

**Comment by jpivarski** [comment]
PR #131 took another bite out of this issue, but didn't finish everything.
- the design note with the current-state audit, manifest contents, dropped upstream artifacts, candidate feature table, path notes, regeneration commands, and an alignment verdict in [docs/deepsets_issue119/README.md](/Users/jpivarski/dsi/vanguard/docs/deepsets_issue119/README.md:15)
- the alignment script with CLI support for config path, DCE root, vessel-seg root, case IDs, time index, slice override, color mode, and output dir in [scripts/deepsets_alignment_check.py](/Users/jpivarski/dsi/vanguard/scripts/deepsets_alignment_check.py:317)
- the four committed overlay PNGs under `docs/deepsets_issue119/figures/`
